tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ConcreteFunction.java

```diff
@@ -154,78 +178,191 @@ public Signature signature() {
-    final Session.Runner runner = session.runner();
+  /**
+   * Get the {@link FunctionDef} proto.
+   */
+  public FunctionDef getFunctionDef() {
+    return nativeFunction.getFunctionDef();
+  }
 
-    signatureDef.getInputsMap().forEach((argName, t) -> {
-      Tensor tensor = arguments.get(argName);
-      if (tensor == null) {
-        throw new IllegalArgumentException(String.format("Missing argument [%s]", argName));
-      }
-      runner.feed(t.getName(), tensor);
-    });
+  /**
+   * Get whether the function is stateful.
+   */
+  public boolean isStateful() {
+    return nativeFunction.isStateful();
+  }
 
-    Map<String, TensorInfo> outputToNode = signatureDef.getOutputsMap();
-    outputToNode.values().forEach(t -> runner.fetch(t.getName()));
+  Set<TF_Function> getDependencies() {
+    return dependencies;
+  }
 
-    List<Tensor> resultTensors = runner.run();
-    try {
-      ListIterator<Tensor> resultTensorIter = resultTensors.listIterator();
-      Map<String, Tensor> returnMap = new HashMap<String, Tensor>();
+  @Override
+  public void close() {
+    scope.close();
+  }
 
-      // Use the output names as present in the signature definition
-      for (String nodeName: outputToNode.keySet()) {
-        returnMap.put(nodeName, resultTensorIter.next());
+  @Override
+  public String toString() {
+    return signature.toString();
+  }
+
+  public static final String CALL_OP = "PartitionedCall";
+  public static final String STATEFUL_CALL_OP = "StatefulPartitionedCall";
+
+  public Map<String, Operand<?>> call(Scope scope,
+      Map<String, Operand<?>> arguments) {
+    List<Operand<?>> inputList = new ArrayList<>();
+
+    for (String inputName : signature().inputNames()) {
+      Operand<?> input = arguments.get(inputName);
+      if (input == null) {
+        throw new IllegalArgumentException(
+            "Function " + signature().methodName() + " has parameter \"" + inputName
+                + "\", but no argument was passed for it.");
       }
-      return returnMap;
+      inputList.add(input);
+    }
+
+    scope.env().attachFunction(this);
+    String name = getNativeFunctionName();
+
+    String displayName = Scope.isValidOpName(name) ? name : "FunctionCall";
+
+    OperationBuilder opBuilder = scope.env()
+        .opBuilder(isStateful() ? STATEFUL_CALL_OP : CALL_OP, scope.makeOpName(displayName));
+
+    opBuilder.addInputList(inputList.stream().map(Operand::asOutput).toArray(Output[]::new));
+
+    opBuilder.setAttr("f", this);
+    opBuilder.setAttr("Tin", inputList.stream().map(x -> x.asOutput().dataType()).toArray(DataType[]::new));
+    opBuilder.setAttr("Tout", signature().getOutputs().values().stream().map(x -> x.dataType).toArray(DataType[]::new));
+
+    opBuilder = scope.apply(opBuilder);
+    Operation op = opBuilder.build();
+
+    int numOutputs1 = op.numOutputs();
+    List<Operand<?>> outputList = new ArrayList<>(signature().outputNames().size());
```

Comment:
same thing, I don't think we need to collect `outputList` first, we can simply access `op.output(i)` when iterating through the list of `outputNames` below.